torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,135 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
       false, "This codegen backend does not implement call_with_numel");
 }
 
-StmtPtr insertAllocFree(std::unordered_set<BufPtr>& interm_bufs, StmtPtr stmt) {
+c10::optional<size_t> bufSize(BufPtr buf) {
+  size_t size = elementSize(buf->dtype().scalar_type()) * buf->dtype().lanes();
+  for (auto& d : buf->dims()) {
+    if (!d->isConstant()) {
+      return c10::nullopt;
+    }
+    size = size * (*intValue(d));
+  }
+  return size;
+}
+
+// This algorithm takes the list of intermediate buffers and their liveness
+// ragnes, and returns a map of these buffers to memory blocks. A memory block
+// is indicated by the buffer for which it is allocated. Specifically, we
+// linearly scan the intermediate buffers by the time they appear, and try to
+// assign it an existing non-occupied memory block. If there are no such memory
+// blocks available, we'll create memory for it. Once we are beyond the liveness
+// range of this buffer, we'll mark its corresponding memory block as "up for
+// grabs" for future reuse.
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    const std::unordered_set<BufPtr>& bufs,
+    const std::unordered_map<BufPtr, std::tuple<int32_t, int32_t>>&
+        buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    return std::get<0>(buf_ranges.at(b1)) < std::get<0>(buf_ranges.at(b2));
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  // Map intermediate buffers to the most recently used memory if any.
+  std::list<BufPtr> mem_up_for_grabs;
+  std::unordered_map<BufPtr, BufPtr> curr_buf_mem_map;
+  std::vector<std::pair<BufPtr, BufPtr>> global_buf_mem_map;
```

Comment:
What's the difference between "curr" and "global" here? It seems that the name is misleading and the actual meaning is somewhat different - i.e. one of them is simply a vector while the other is a map. Am I missing something?

Reply:
`curr_buf_mem_map` holds the mapping of alive buffers to memories at the moment; and `global_buf_mem_map` holds all the mappings as a vector so we can insert `Alloc/PlacementAlloc` later with it with determined order. Fixing the names.